text/0000-futures.md

```diff
@@ -0,0 +1,520 @@
+        (&mut *self.executor.optimized_queue.get()).push(self.clone())
+    }
+}
+```
+
+The use of `&Arc<Self>` rather than just `&self` makes it possible to work directly with
+the trait object for `Wake`, including cloning it. With `UnsafeWake` below, we'll see
+an API with greater flexibility for the cases where `Arc` is problematic.
+
+In general async values are not coupled to any particular executor, so we use trait
+objects to handle waking. These come in two forms: `Waker` for the general case, and
+`LocalWaker` to provide more effenciency when the wakeup is guaranteed to occur on the
+executor thread:
+
+```rust
+/// A `Waker` is a handle for waking up a task by notifying its executor that it
+/// is ready to be run.
+///
+/// This handle contains a trait object pointing to an instance of the `UnsafeWake`
+/// trait, allowing notifications to get routed through it.
+///
+/// Implements `Clone`, `Send`, and `Sync`.
+pub struct Waker { ... }
+
+impl Waker {
+    /// Wake up the task associated with this `Waker`.
+    pub fn wake(&self);
+}
+
+/// A `LocalWaker` is a handle for waking up a task by notifying its executor that it is ready to be run.
+///
+/// This is similar to the `Waker` type, but cannot be sent across threads. Task executors can use this type to implement more optimized singlethreaded wakeup behavior.
+impl LocalWaker {
+    /// Wake up the task associated with this `LocalWaker`.
+    pub fn wake(&self);
+}
+
+/// You can upgrade to a sendable `Waker` at zero cost, but waking through a `Waker` is more expensive
+/// due to synchronization.
+impl From<LocalWaker> for Waker  { .. }
+```
+
+Task execution always happens in the context of a `LocalWaker` that can be used to
+wake the task up locally, or converted into a `Waker` that can be sent to other threads.
+
+It's possible to construct a `Waker` using `From<Arc<dyn Wake>>`.
+
+### `UnsafeWake` and `no_std` compatibility
+
+The [`UnsafeWake` trait](https://doc.rust-lang.org/nightly/std/task/trait.UnsafeWake.html)
+in `core::task` is designed to support task wakeup in  `no_std` environments, where
+we cannot use `Arc`. It is *not* proposed for stabilization at this time, because
+its APIs are awaiting revision based on object safety for `*mut self` methods.
+
+## `core::future` module
+
+With all of the above task infrastructure in place, defining `Future` is
+straightforward:
+
+```rust
+pub trait Future {
+    /// The type of value produced on completion.
+    type Output;
+
+    /// Attempt to resolve the future to a final value, registering
+    /// the current task for wakeup if the value is not yet available.
+    ///
+    /// # Return value
+    ///
+    /// This function returns:
+    ///
+    /// - [`Poll::Pending`] if the future is not ready yet
+    /// - [`Poll::Ready(val)`] with the result `val` of this future if it
+    ///   finished successfully.
+    ///
+    /// Once a future has finished, clients should not `poll` it again.
```

Comment:
No behavior is specified for when clients do do that. I think we should say something. For example, "implementors may panic".

Reply:
I think we could even go a bit stronger than that, “implementors should panic, but clients may not rely on this this”. All `async fn` futures guarantee this, and I believe so do the current `futures 0.3` adaptors.

I think it would also be good to mention that calling `poll` again must not cause memory unsafety. The current mention that it can do anything at all makes it seem like it is allowed to have undefined behaviour, but since this is not an `unsafe fn` the implementer cannot rely on the client’s behaviour for memory safety purposes.